unicodetools/src/main/java/org/unicode/text/UCD/GenerateData.java

```diff
@@ -938,11 +945,260 @@ public static void writeNormalizerTestSuite(String directory, String fileName)
         }
 
         Utility.fixDot();
+
+        System.out.println("Writing Part 4");
+        log.println("#");
+        log.println("@Part4 # Canonical closures (excluding Hangul)");
+        log.println("#");
+        for (var entry : canonicalDecompositionsByCodepoint.entrySet()) {
+            final int cp = entry.getKey();
+            final String decomposition = entry.getValue();
+            if (cp >= 0xAC00 && cp <= 0xD7A3) {
+                continue;
+            }
+            forAllStringsCanonicallyDecomposingTo(
+                    decomposition,
+                    s -> {
+                        // Skip NFD and single code points (NFC or full
+                        // composition exclusions), already covered in Part 1.
+                        if (!s.equals(decomposition) && s.codePointCount(0, s.length()) != 1) {
+                            writeLine(s, log, true);
+                        }
+                    });
+        }
+
+        System.out.println("Writing Part 5");
+        log.println("#");
+        log.println("@Part5 # Chained primary composites");
+        log.println("#");
+
+        // Mutable maps because we end up adding some C↦C mappings for
+        // convenience below.
+        final Map<Integer, Set<Integer>> primaryCompositesByFirstNFDCodePoint = new TreeMap<>();
+        final Map<Integer, Set<Integer>> primaryCompositesByLastNFDCodePoint = new TreeMap<>();
+        for (var entry : canonicalDecompositionsByCodepoint.entrySet()) {
+            final int cp = entry.getKey();
+            final String decomposition = entry.getValue();
+            if (Default.nfc().normalize(cp).equals(Character.toString(cp))) {
+                int first = decomposition.codePointAt(0);
+                int last = decomposition.codePointBefore(decomposition.length());
+                primaryCompositesByFirstNFDCodePoint
+                        .computeIfAbsent(first, key -> new TreeSet<>())
+                        .add(cp);
+                primaryCompositesByLastNFDCodePoint
+                        .computeIfAbsent(last, key -> new TreeSet<>())
+                        .add(cp);
+            }
+        }
+
+        Collection<String> skippedNFCs = new ArrayList<>();
+        Set<String> part5NFCs = new TreeSet<>();
+
+        // The set of all sequences of two code points appearing within a
+        // canonical decomposition.
+        Set<String> links = new TreeSet<>();
+        for (String decomposition : canonicalDecompositionsOfSingleCodepoints) {
+            int first = decomposition.codePointAt(0);
+            int second;
+            for (int i = UTF16.getCharCount(first);
+                    i < decomposition.length();
+                    first = second, i += UTF16.getCharCount(first)) {
+                second = decomposition.codePointAt(i);
+                links.add(Character.toString(first) + Character.toString(second));
```

Comment:
optional, maybe wrong:

It seems like you could populate the primaryCompositesBy(First,Last)NFDCodePoint maps directly here, without collecting a list of links...??

Reply:
It has been populated already, except for the cheesy self-mappings added lazily below.